openupgrade_scripts/scripts/product/15.0.1.2/pre-migration.py

```diff
@@ -0,0 +1,55 @@
+from openupgradelib import openupgrade
+
+
+def compute_product_template_attribute_line_value_count(env):
+    if not openupgrade.column_exists(
+        env.cr, "product_template_attribute_line", "value_count"
+    ):
+        openupgrade.logged_query(
+            env.cr,
+            """
+            ALTER TABLE product_template_attribute_line
+            ADD COLUMN value_count integer
+            """,
+        )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE product_template_attribute_line al
+        SET value_count = (SELECT COUNT(product_template_attribute_line_id)
+        FROM   product_attribute_value_product_template_attribute_line_rel
+        WHERE  product_template_attribute_line_id = al.id)
+        """,
+    )
+
+
+def convert_text_to_html(env):
+    openupgrade.convert_field_to_html(
+        env.cr, "product_template", "description", "description"
+    )
+
+
+def compute_detailed_type_and_type(env):
+    if not openupgrade.column_exists(env.cr, "product_template", "detailed_type"):
+        openupgrade.logged_query(
```

Comment:
Is there any reason we cannot use rename column?

Reply:
Oh, the old `type` column is still there as well. The new field `detailed_type` can branch out to things like 'event' which are then automatically mapped to value `service` for `type`.

Reply:
@sang250399 I do wonder what the use case is of a defensive strategy like this one. If the field already exists, what are the odds that the values are compatible with what is introduced in Odoo 15? Would it not be safer to create an unconditional copy and let OpenUpgrade halt on the SQL error if the column already exists?

Reply:
@StefanRijnhart I don't believe we can be certain whether the column is already present. As a result, this sql check simply prevents an error from occurring because the column already exists. If it already exists and contains data, replace it with a new value that corresponds to code 15. (the old data of that column may not be correct)
cc: @sang250399

Reply:
Yes alright. It's a corner case anyway and it's no different from Odoo introducing new columns that may have already been claimed by third parties in earlier editions.